include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td

```diff
@@ -144,4 +144,64 @@ class IsValidAIE2MatMulShapeAndType<string lhs, string rhs, string acc> :
                                    rhs, VectorOfShapeAndType<[8, 4], BF16>,
                                    acc, VectorOfShapeAndType<[4, 4], F32>>]>>;
 
+def AIE2MulElemLHS :
+  AnyTypeOf<[VectorOfShapeAndType<[32], I8>,
+             VectorOfShapeAndType<[32], I16>,
+             VectorOfShapeAndType<[16], I32>,
+             VectorOfShapeAndType<[16], BF16>,
+             VectorOfShapeAndType<[16], F32>], 
+            "a vector compatible with a lhs operand of element-wise multiply and "
+            # "accumulate",
+            "::mlir::VectorType">;
+
+def AIE2MulElemRHS :
+  AnyTypeOf<[VectorOfShapeAndType<[32], I8>,
+             VectorOfShapeAndType<[32], I16>,
+             VectorOfShapeAndType<[16], I32>,
+             VectorOfShapeAndType<[16], BF16>,
+             VectorOfShapeAndType<[16], F32>],
+            "a vector compatible with a rhs operand of element-wise multiply and "
+            # "accumulate",
+            "::mlir::VectorType">;
```

Comment:
This is overkill. Lhs & Rhs types have to match, you don't need to spell out every single valid type for each of the operands.

Reply:
The op verifier checks that the lanes and datatypes are the same for operands. See https://github.com/Xilinx/mlir-aie/blob/aba1887b9750dea8db14fbe16417068f83a6b94c/lib/Dialect/AIEVec/IR/AIEVecOps.cpp#L872-L894

This can probably be replaced by `SameTypeOperands` and `SameOperandsAndResultShape` traits....

Reply:
However, I'm not sure how to better constraint the fact that the `acc` is i32 for i8/i16 operands, `i64` for i32 operands, and `f32` for bf16/f32 operands. Maybe just constraints it in the op verifier in cpp code?

Reply:
> The op verifier checks that the lanes and datatypes are the same for operands. See
> 
> https://github.com/Xilinx/mlir-aie/blob/aba1887b9750dea8db14fbe16417068f83a6b94c/lib/Dialect/AIEVec/IR/AIEVecOps.cpp#L872-L894
> 

Notice that the C++ verifier will be replaced by the automatically generated one.

> This can probably be replaced by `SameTypeOperands` and `SameOperandsAndResultShape` traits....

Indeed.

Reply:
> However, I'm not sure how to better constraint the fact that the `acc` is i32 for i8/i16 operands, `i64` for i32 operands, and `f32` for bf16/f32 operands. Maybe just constraints it in the op verifier in cpp code?

For that, you can either define a new predicate ('IsValidAccumulatorTypeFor`) that verifies the accumulator type validity using simpler pre-existing predicates, or define a C++ function that makes that check and invoke it from the predicate.